### PR TITLE
Ensure that NI_Vector_Dot is always handled as TYP_SIMD

### LINKED ISSUE #83316: Test failure System.Numerics.Tests.Matrix4x4Tests.Matrix4x4CreateShadowTest02
Failed in Run: [runtime-coreclr libraries-jitstress2-jitstressregs 20230311.1](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_build/results?buildId=201383)
 
 **Failed tests:**
 ```
 net8.0-windows-Release-x64-CoreCLR_checked-jitstress2_jitstressregs3-Windows.10.Amd64.Open
     - System.Numerics.Tests.Matrix4x4Tests.Matrix4x4CreateShadowTest02
 ```
 
 **Error message:**
 ```
 Matrix4x4.CreateShadow did not provide expected value.
Expected: True
Actual:   False
 ```
 
 **Stack trace:**
 ```
   at System.Numerics.Tests.Matrix4x4Tests.Matrix4x4CreateShadowTest02() in /_/src/libraries/System.Numerics.Vectors/tests/Matrix4x4Tests.cs:line 762
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr) in /_/src/libraries/System.Private.CoreLib/src/System/Reflection/MethodInvoker.cs:line 59
 ```

### PR DESCRIPTION
**Pull request by tannergooding**
This resolves https://github.com/dotnet/runtime/issues/83316

### CODE REVIEW

File: src/coreclr/jit/morph.cpp [hunk 1]
```diff
@@ -10670,39 +10670,45 @@ GenTree* Compiler::fgOptimizeHWIntrinsic(GenTreeHWIntrinsic* node)
...
+            // then look for four common patterns:
             //  * Create(Dot(..., ...))
             //  * Create(Sqrt(Dot(..., ...)))
+            //  * Create(ToScalar(Dot(..., ...)))
+            //  * Create(ToScalar(Sqrt(Dot(..., ...))))
             //
-            // When these exist, we'll avoid converting to a scalar at all and just
-            // keep everything as a vector. However, we only do this for Vector64/Vector128
-            // and only for float/double.
+            // When these exist, we'll avoid converting to a scalar and avoid broadcasting
```
Comment (kunalspathak) [inline src/coreclr/jit/morph.cpp:10688]:
```suggestion
            // When these exist, we'll avoid converting to a scalar and hence, avoid broadcasting
```

File: src/coreclr/jit/morph.cpp [hunk 2]
```diff
@@ -10670,39 +10670,45 @@ GenTree* Compiler::fgOptimizeHWIntrinsic(GenTreeHWIntrinsic* node)
...
-                break;
-            }
-
-            GenTree* op1  = node->Op(1);
-            GenTree* sqrt = nullptr;
+            GenTree* op1      = node->Op(1);
+            GenTree* sqrt     = nullptr;
+            GenTree* toScalar = nullptr;
 
             if (op1->OperIs(GT_INTRINSIC))
```
Comment (kunalspathak) [inline src/coreclr/jit/morph.cpp:10705]: Wondering what are the examples where we now want to have `varTypeIsFloating()` check inside the `OperIs(GT_INTRINSIC)` check?

File: src/coreclr/jit/morph.cpp [hunk 3]
```diff
@@ -10719,6 +10725,24 @@ GenTree* Compiler::fgOptimizeHWIntrinsic(GenTreeHWIntrinsic* node)
...
+#endif
+            {
+                op1 = hwop1->Op(1);
+
+                if (!op1->OperIs(GT_HWINTRINSIC))
+                {
+                    break;
+                }
+
+                toScalar = hwop1;
```
Comment (kunalspathak) [inline src/coreclr/jit/morph.cpp:10742]: assert that `toScalar` is indeed `ToScalar`?

File: src/coreclr/jit/morph.cpp (hunk 2 above)
Comment (tannergooding) [reply src/coreclr/jit/morph.cpp:10705]: The codegen for integers was already doing the "right thing" and producing vectors where the result is in all elements, so the `varTypeIsFloating` check was unnecessary. It's only needed in the case we have a `GT_INTRINSIC` in-between, in which case we can only handle `Sqrt` and only for floating-point

File: src/coreclr/jit/morph.cpp (hunk 3 above)
Comment (tannergooding) [reply src/coreclr/jit/morph.cpp:10742]: We're in a very small `if condition` that can only be `ToScalar`. The asserts would basically duplicate the larger `#ifdef` here, which is why I avoided it.

File: src/coreclr/jit/morph.cpp [hunk 4]
```diff
@@ -10719,6 +10725,24 @@ GenTree* Compiler::fgOptimizeHWIntrinsic(GenTreeHWIntrinsic* node)
 
             GenTreeHWIntrinsic* hwop1 = op1->AsHWIntrinsic();
 
+#if defined(TARGET_ARM64)
+            if ((hwop1->GetHWIntrinsicId() == NI_Vector64_ToScalar) &&
+                (hwop1->GetHWIntrinsicId() == NI_Vector128_ToScalar))
```
Comment (tannergooding) [inline src/coreclr/jit/morph.cpp:10730]:
```suggestion
            if ((hwop1->GetHWIntrinsicId() == NI_Vector64_ToScalar) ||
                (hwop1->GetHWIntrinsicId() == NI_Vector128_ToScalar))
```

File: src/coreclr/jit/morph.cpp (hunk 3 above)
Comment (kunalspathak) [reply src/coreclr/jit/morph.cpp:10742]: ah yes, i forgot about it.

**Comment by tannergooding** [comment]
@kunalspathak, any other feedback or is this good to merge?

**Review by kunalspathak — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**